### DISCUSSION TITLE
How do I reload my plugin without having to restart Binary ninja.

### DISCUSSION THREAD

**Discussion by deqline** [Vector35/binaryninja-api#1734] [General] [answered]
Question above.

**Comment by psifertex** [comment, marked as answer]
First, the bad news. If you're writing an architecture plugin or a binaryview plugin there is no way to reload the plugin without restarting BN at this time. That said, there is at least a few things that can help even with that workflow.

First, you can bind a hotkey to the Restart Binary Ninja action (customize hotkeys via CMD/CTL+SHIFT+b). Next, if you launch BN from the command-line with a file-parameter and then use that restart option, it'll automatically re-open the same file. This makes prototyping architecture and views much easier.

Alternatively, if you have a commercial license, you can simply write a stand-alone command-line plugin that dumps some disassembly from a function and you can run that after making changes to an architecture or view plugin to have it refresh the results as-needed.

Ok, now for the better news. If you're writing any other kind of plugin, here's the current recommended workflow:

1) Put your plugin in your [plugins folder](https://docs.binary.ninja/getting-started.html#user-folder). (You can also get to it from the Tools menu). For example: `~/.binaryninja/plugins/modulename.py`
2) Do NOT use any of the typical [register_for](https://api.binary.ninja/search.html?q=register_for&check_keywords=yes&area=default) APIs that you would normally use in a plugin. Or you can write it, but just comment it out temporarily while testing your plugin.
3) Next, manually load your plugin into the scripting console (and load importlib too as we're going to need it shortly):
```
import modulename 
import importlib
```
4) Now when you want to test the function that would have been called as a callback from the `register_for` function above, just recreate it in the console and put a reload before it so every time you run this command it will reload your plugin and try the latest version:
```
importlib.reload(modulename) #Note this takes an actual module not a string of the name
modulename.handle_address(be, here)
```
5) Next, make any changes you want to the plugin, then just hit up, enter in the console to re-run the last command. You could even use a [snippet](https://github.com/Vector35/snippets ) to trigger it via a hotkey instead.
7) When you're done developing your plugin, just re-enable the `register_for` api to call your callback for you and the plugin will be available from all of the appropriate menus that you registered for.

**Reply by deqline** [reply]
Perfect, Thank you! I will try to do that :D

**Reply by deqline** [reply]
One more question though @psifertex where can I find my module name? is it in plugin.json or is it my folder name? When I try to import it I get a syntax error.

**Reply by psifertex** [reply]
That's the name of the file you create in the plugin folder.

**Reply by deqline** [reply]
Oh right, turns out I typed it correctly but It didn't accept the `-` character. I had to rename it.

**Reply by deqline** [reply]
@psifertex So I did everything and now I can directly import it from the console. Now I do have a problem; when I try to do:
```
importlib.reload(my_module_name);
my_module_name.main(bv, here)
```
I get an error saying: `Failed connecting to discord RPC There is no current event loop in thread 'Thread-1'`
Now is that a problem with bn api or discord rpc?

**Reply by saagarjha** [reply]
fyi this should help you if you don't want to rename your module:

```python
plugin = importlib.import_module("my-module-name")
importlib.reload(plugin);
```
